File: python/ONI_capture.py
from datetime import datetime
import numpy as np
import time
import os
import cv2

from glob import glob
from os.path import join
from openni import openni2
from openni import _openni2 as c_api
import openni

width = 640
height = 480

def colormap(depth):
    # Apply colormap
    depth[depth > 10000.] = 0
    depthColored = cv2.convertScaleAbs(depth, alpha=(255.0/10000.0))
    depthColored = cv2.applyColorMap(depthColored, cv2.COLORMAP_JET)
    depthColored[depth==0] = np.array([0,0,0])
    return depthColored


if __name__ == "__main__":

    save = False

    # Set the path to the orbbec SDK openni lib
    libpath = "D:/Orbbec/Tools/Windows_V2.3.0.66/windows/SDK/x64/Redist"
    openni2.initialize(libpath)
    dev = openni2.Device.open_any()
    print(dev.get_device_info())

    dev.set_depth_color_sync_enabled(True)
    # fail to set registration mode
    # dev.set_image_registration_mode(openni2.IMAGE_REGISTRATION_DEPTH_TO_COLOR)

    depth_stream = dev.create_depth_stream()
    depth_stream.start()

    color_capture = cv2.VideoCapture(0)
    print("Backend: ", color_capture.get(cv2.CAP_PROP_BACKEND))

    if save:
        outDir = join("../", str(time.time()))
        os.mkdir(outDir)
        os.mkdir(join(outDir, 'rgb'))
        os.mkdir(join(outDir, 'depth'))
        
        frameCount = 0
        timestampList = []  

    while True:

        # Read frame from buffer
        print("Depth Timestamp: ", time.time())
        depthBuf = depth_stream.read_frame().get_buffer_as_uint16()
        print("Color Timestamp: ", time.time())
        ret, color = color_capture.read()

        # Parser depth frame
        depth = np.frombuffer(depthBuf, dtype=np.uint16).reshape(height, width) 
        depth = cv2.flip(depth, 1)
        
        if save:
            # Save Image
            timestamp = time.time()
            depthFilename = os.path.join(outDir, "depth", "{:.06f}.png".format(timestamp))
            colorFilename = os.path.join(outDir, "rgb", "{:.06f}.png".format(timestamp))
            cv2.imwrite(depthFilename, depth.astype(np.uint16))
            cv2.imwrite(colorFilename, color)
            timestampList.append(timestamp)
            print("Frame Saved: {:06d}".format(frameCount))
            frameCount += 1
        
        # Display
        cv2.imshow('depth', colormap(depth))
        cv2.imshow('color', color)
        overlay = cv2.addWeighted(color, 0.7, colormap(depth), 0.5, 0)
        cv2.imshow('Rectified', overlay)        

        key = cv2.waitKey(10) & 0xFF
        if key == 27: # Esc
            break

    if save:
        # Create depth.txt & rgb.txt  
        depthFile = open(join(outDir, "depth.txt"), 'w')
        rgbFile = open(join(outDir, "rgb.txt"), 'w')
        associationFile = open(join(outDir, "association.txt"), 'w')

        for timestamp in timestampList:

            depthPath = "depth/{:.06f}.png".format(timestamp)
            rgbPath = "rgb/{:.06f}.png".format(timestamp)

            depthFile.write("{:.06f} {}\n".format(timestamp, depthPath))
            rgbFile.write("{:.06f} {}\n".format(timestamp, rgbPath))
            associationFile.write("{:.06f} {} {:.06f} {}\n".format(timestamp, depthPath, timestamp, rgbPath))

        depthFile.close()
        rgbFile.close()
        associationFile.close()

    depth_stream.close()
    color_capture.release()
    dev.close()

File: cpp/ONI_capture.cpp
// ONI_capture.cpp : Defines the entry point for the application.
//

#include <stdlib.h>
#include <iostream>
#include <fstream>
#include <sstream>
#include <iomanip>
#include <string>
#include <direct.h>
#include <ctime>
#include <Windows.h>

#include "OpenNI.h"
#include "opencv2/core/core.hpp"
#include "opencv2/highgui/highgui.hpp"
#include "opencv2/imgproc/imgproc_c.h"
#include "opencv2/imgproc/imgproc.hpp"
#include "opencv2/calib3d/calib3d.hpp"

using namespace std;

#define SAMPLE_READ_WAIT_TIMEOUT 5000 //2000ms


int main(int argc, char** argv)
{

	bool saveFrame = true;

	std::string outDir = "..\\CapturedSequence\\";
	std::time_t timestamp = std::time(0);
	std::stringstream ss;
	ss << timestamp;
	std::string ts = ss.str();
	std::string outputFolder = std::string(outDir + ts);

	std::string depthFolder = std::string(outputFolder + "\\depth\\");
	std::string colorFolder = std::string(outputFolder + "\\color\\");

	if (saveFrame) {

		if ((CreateDirectory(outputFolder.c_str(), NULL)) &&
			(CreateDirectory(depthFolder.c_str(), NULL) || ERROR_ALREADY_EXISTS == GetLastError()) &&
			(CreateDirectory(colorFolder.c_str(), NULL) || ERROR_ALREADY_EXISTS == GetLastError()))
		{
			cout << "Saving frame to: " << outputFolder << "\n";
		}
		else {
			cout << "Fail to create output folder: " << outputFolder << "\n";
		}
	}

	openni::Status result = openni::STATUS_OK;

	//OpenNI2 Frame
	openni::VideoFrameRef oniDepthImg;
	cv::Mat mImageDepth_colormap, mImageDepth_scale, mImageColor, mImageOverlay;

	char key = 0;

	// Set up openni
	result = openni::OpenNI::initialize();
	if (result != openni::STATUS_OK)
	{
		printf("Initialize failed\n%s\n", openni::OpenNI::getExtendedError());
		return 1;
	}

	// Open camera device
	openni::Array<openni::DeviceInfo> deviceList;
	openni::OpenNI::enumerateDevices(&deviceList);
	int cc = deviceList.getSize();
	if (cc <= 0)
	{
		cout << "No device found\n";
		return -1;
	}
	const char* deviceUri;
	deviceUri = deviceList[0].getUri();
	cout << "Device URI: " << deviceUri << "\n";

	openni::Device device;
	result = device.open(deviceUri);

	// set depth video mode
	openni::VideoMode modeDepth;
	modeDepth.setResolution(640, 480);
	modeDepth.setFps(30);
	modeDepth.setPixelFormat(openni::PIXEL_FORMAT_DEPTH_1_MM);

	// create depth stream
	openni::VideoStream oniDepthStream;
	oniDepthStream.setVideoMode(modeDepth);
	oniDepthStream.create(device, openni::SENSOR_DEPTH);
	int changedStreamDummy;
	openni::VideoStream* pStream = &oniDepthStream;

	if (device.isImageRegistrationModeSupported(openni::IMAGE_REGISTRATION_DEPTH_TO_COLOR))
	{
		device.setImageRegistrationMode(openni::IMAGE_REGISTRATION_DEPTH_TO_COLOR);
	}
	device.setDepthColorSyncEnabled(true);

	float vfov = oniDepthStream.getVerticalFieldOfView();
	float hfov = oniDepthStream.getHorizontalFieldOfView();

	cout << "FOV: " << vfov << " : " << hfov << "\n";

	// start depth stream
	result = oniDepthStream.start();  
	if (result != openni::STATUS_OK) {
		cout << "Fail to open depth stream\n";
	}

	// start color stream
	cv::VideoCapture colorStream(0, cv::CAP_DSHOW);
	if (!colorStream.isOpened()) {
		cout << "Fail to open color stream\n";
	}
	colorStream.set(cv::CAP_PROP_BUFFERSIZE, 3);
	
	cout << "FPS: " << colorStream.get(cv::CAP_PROP_FPS) << "\n";
	cout << "Buffer Size: " << colorStream.get(cv::CAP_PROP_BUFFERSIZE) << "\n";
	cout << "SAR: " << colorStream.get(cv::CAP_PROP_SAR_NUM) << "\n";

	string depthWinName = "Depth";
	string colorWinName = "Color";
	string overlayWinName = "Overlay";

	cv::namedWindow(depthWinName, 1);
	cv::namedWindow(colorWinName, 1);
	cv::namedWindow(overlayWinName, 1);

	int frameCount = 0;

	while ((key != 27)) // ESC
	{
		openni::OpenNI::waitForAnyStream(&pStream, 1, &changedStreamDummy, SAMPLE_READ_WAIT_TIMEOUT);
		oniDepthStream.readFrame(&oniDepthImg);
		colorStream >> mImageColor;

		// Parser depth
		const cv::Mat mImageDepth(oniDepthImg.getHeight(), oniDepthImg.getWidth(), CV_16UC1, (void*)oniDepthImg.getData());
		cv::flip(mImageDepth, mImageDepth, 1);
		mImageDepth.convertTo(mImageDepth_scale, CV_8U, 255.0 / 8000.0);
		cv::threshold(mImageDepth_scale, mImageDepth_scale, 8000.0, 8000.0, cv::THRESH_TOZERO_INV);
		cv::applyColorMap(mImageDepth_scale, mImageDepth_colormap, cv::COLORMAP_JET);
		
		for (int v = 0; v < mImageDepth_colormap.rows; v++){
			uchar* p_colormap = mImageDepth_colormap.ptr<uchar>(v);
			uchar* p_scale= mImageDepth_scale.ptr<uchar>(v);
			for (int u = 0; u < mImageDepth_colormap.cols; u++) {
				if (p_scale[u] == 0) {
					p_colormap[u * 3 + 0] = 0;
					p_colormap[u * 3 + 1] = 0;
					p_colormap[u * 3 + 2] = 0;
				}
			}
		}
		
		// Show img
		cv::imshow(depthWinName, mImageDepth_colormap);
		cv::imshow(colorWinName, mImageColor);

		cv::addWeighted(mImageColor, 0.7, mImageDepth_colormap, 0.5, 0, mImageOverlay);
		cv::imshow(overlayWinName, mImageOverlay);

		// Save frame
		if (saveFrame) {
			std::stringstream fs;
			fs << std::setfill('0') << std::setw(6) << frameCount;
			std::string countStr = fs.str();
			std::string depthFileName = std::string(depthFolder + countStr + ".depth.png");
			std::string colorFileName = std::string(colorFolder + countStr + ".color.jpg");

			cv::imwrite(depthFileName, mImageDepth);
			cv::imwrite(colorFileName, mImageColor);
		}

		key = cv::waitKey(1);
		frameCount++;

	}

	//destroy
	oniDepthStream.destroy(); 
	device.close();
	openni::OpenNI::shutdown();
	colorStream.release();
	return 0;
}